transaction/transaction.go

```diff
@@ -806,3 +807,117 @@ func byte32FromBase64(in string) (out [32]byte, err error) {
+func MakeUnsignedAppOptInTx(appIdx uint64, appArgs [][]byte, accounts []string, foreignApps []uint64) (tx types.Transaction, err error) {
+	return MakeUnsignedApplicationCallTx(appIdx, appArgs, accounts, foreignApps, types.OptInOC, nil, nil, emptySchema, emptySchema)
+}
+
+// MakeUnsignedAppCloseOutTx makes a transaction for closing out of
+// (deallocating all account-specific state for) an application
+func MakeUnsignedAppCloseOutTx(appIdx uint64, appArgs [][]byte, accounts []string, foreignApps []uint64) (tx types.Transaction, err error) {
+	return MakeUnsignedApplicationCallTx(appIdx, appArgs, accounts, foreignApps, types.CloseOutOC, nil, nil, emptySchema, emptySchema)
+}
+
+// MakeUnsignedAppClearStateTx makes a transaction for clearing out all
+// account-specific state for an application. It may not be rejected by the
+// application's logic.
+func MakeUnsignedAppClearStateTx(appIdx uint64, appArgs [][]byte, accounts []string, foreignApps []uint64) (tx types.Transaction, err error) {
+	return MakeUnsignedApplicationCallTx(appIdx, appArgs, accounts, foreignApps, types.ClearStateOC, nil, nil, emptySchema, emptySchema)
+}
+
+// MakeUnsignedAppNoOpTx makes a transaction for interacting with an existing
+// application, potentially updating any account-specific local state and
+// global state associated with it.
+func MakeUnsignedAppNoOpTx(appIdx uint64, appArgs [][]byte, accounts []string, foreignApps []uint64) (tx types.Transaction, err error) {
+	return MakeUnsignedApplicationCallTx(appIdx, appArgs, accounts, foreignApps, types.NoOpOC, nil, nil, emptySchema, emptySchema)
+}
+
+// MakeUnsignedApplicationCallTx is a helper for the above ApplicationCall
+// transaction constructors. A fully custom ApplicationCall transaction may
+// be constructed using this method.
+func MakeUnsignedApplicationCallTx(appIdx uint64, appArgs [][]byte, accounts []string, foreignApps []uint64, onCompletion types.OnCompletion, approvalProg []byte, clearProg []byte, globalSchema types.StateSchema, localSchema types.StateSchema) (tx types.Transaction, err error) {
+	tx.Type = types.ApplicationCallTx
+	tx.ApplicationID = types.AppIndex(appIdx)
+	tx.OnCompletion = onCompletion
+
+	tx.ApplicationArgs = appArgs
+	tx.Accounts, err = parseTxnAccounts(accounts)
+	if err != nil {
+		return tx, err
+	}
+
+	tx.ForeignApps = parseTxnForeignApps(foreignApps)
+	tx.ApprovalProgram = approvalProg
+	tx.ClearStateProgram = clearProg
+	tx.LocalStateSchema = localSchema
+	tx.GlobalStateSchema = globalSchema
+
+	return tx, nil
+}
+
+func parseTxnAccounts(accounts []string) (parsed []types.Address, err error) {
+	for _, acct := range accounts {
+		addr, err := types.DecodeAddress(acct)
+		if err != nil {
+			return nil, err
+		}
+		parsed = append(parsed, addr)
+	}
+	return
+}
+
+func parseTxnForeignApps(foreignApps []uint64) (parsed []types.AppIndex) {
+	for _, aidx := range foreignApps {
+		parsed = append(parsed, types.AppIndex(aidx))
+	}
+	return
+}
+
+func SetHeader(
```

Comment:
Rephrasing feedback per our call -
- Recommending changing `SetHeader` to use `types.SuggestedParams` for things like the `fee`
- This pattern of using `MakeFoo` paired with `SetHeader` is a new pattern to this sdk, so this likely merits extra documentation and examples in `README.md`
- It is an open design decision whether these new functions should be in `future` or `transaction` - most existing `transaction` functions are deprecated. After talking it out on the phone, I think leaving it in `transaction` makes sense, because this new code does not break backwards compatibility. (`future` was intended only for functions that would break backwards compatibility.)

Reply:
Here is my understanding of SuggestedParams:
- SuggestedParams overlaps with Header
- SuggestedParams is a convenience utility of the client via BuildSuggestedParams
- SuggestedParams struct is used for copying its values over to Header.

My argument is:
1. SuggestedParams should only be involved when BuildSuggestedParams is used. Without BuildSuggestedParams, it is a redundant copy of Header. 

2. SuggestedParams should not be in competition with Header. It is confusing for a user do decide which one to use. 

For example, in unit tests, where no client exists, there should be no SuggestedParams involved. Only Header and other basic structures should be involved.

Therefore, SuggestedParams should not be part of the application building function interface.

Reply:
:+1: Okay! In that case, please put extra explanation and examples in `README.md` so users and developer relations understand this new way of building transactions. You could also add `SetHeaderFromSuggestedParams`, because user devs seem to really like using `types.SuggestedParams` to fill in header elements, but maybe that's messy/confusing.

Reply:
Header already has meaning in the context of an HTTP Client, I'm not sure why we would overload that term in this way. I'm also not sure why we would add another convention for setting these values when we already have 2, and plan on introducing builders as well.

Reply:
> why we would add another convention for setting these values when we already have 2
I think you ask the right question here.
It is because applications transactions already have many fields. Adding on top the basic transaction fields like sender, note... will be too much.

Moreover, the makeTransaction functions do not give a final transaction object. There are examples where other fields are added afterwards. For example: tx.Rekey, tx.Group. 

I will draft another solution, which will replace SetHeader with SetTransactionFields(sg SuggestedParams, sender, note, group, lease, rekeyTo)
This will be a compromise, where the beloved SuggestedParams will be used, without making a freight train out of the makeApplicationTx call.